tools/commands-packages-query.js

```diff
@@ -0,0 +1,1159 @@
+  }
+  // We don't know when the release was published. Luckily, since there is no
+  // way to use the tool outside of a release, and we always change the tool
+  // between releases, it is a good bet that the release was published on the
+  // same day as the tool.
+  var toolPackage = versionRecord.tool.split('@');
+  var toolName = toolPackage[0];
+  var toolVersion = toolPackage[1];
+  var toolRecord = catalog.official.getVersion(toolName, toolVersion);
+  return new Date(toolRecord.published);
+};
+
+// Convert package version dependencies into a more helpful format for output in
+// 'meteor show'.
+var formatDependencies = function (versionDependencies) {
+  var dependencies = _.map(
+    // The dependency on 'meteor' was almost certainly added automatically, by
+    // Isobuild. Returning this to the user will only cause confusion.
+    _.omit(versionDependencies, "meteor"),
+    function (dep, depName) {
+      // We will only consider this a weak dependency if all of its references
+      // are marked as weak.
+      var weak = _.every(dep.references, function (ref) {
+        return !! ref.weak;
+      });
+      return {
+        name: depName,
+        constraint: dep.constraint,
+        weak: weak
+      };
+  });
+  return dependencies;
+};
+
+// Converts an object to an EJSON string with the right spacing.
+var convertToEJSON = function (data) {
+  var EJSON = isopackets.load('ejson').ejson.EJSON;
+  return EJSON.stringify(data, { indent: true }) + "\n";
+};
+
+// Takes in a string and pads it with whitespace to the length of the longest
+// possible date string.
+var padLongformDate = function (dateStr) {
+  var numSpaces = utils.maxDateLength - dateStr.length;
+  return dateStr + Array(numSpaces + 1).join(' ');
+};
+
+// In order to get access to local package data, we need to create a local
+// package catalog. The best way to do that is to create a temporary
+// ProjectContext and let it handle catalog initialization. When we do, we need
+// to make sure that it is aware of all the local packages that we might care
+// about.
+//
+// This function returns such a ProjectContext, and takes in the following
+// options:
+//  - appDir: If we are running in the context of an app, this will contain the
+//    root of the app. We want to make sure to grab the data from the app's
+//    local packages.
+//  - packageDir: If we are running in a package directory, this will contain
+//    the source root of that package. If we are running from inside a package,
+//    we want that package to show up in our results.
+var getTempContext = function (options) {
+  var projectContext;
+  // If we are running in an app, we will use it to create a
+  // (mostly immutable) projectContext.
+  if (options.appDir) {
+    projectContext = new projectContextModule.ProjectContext({
+      projectDir: options.appDir,
+      // Don't change the state of the app.
+      neverWriteProjectConstraintsFile: true
```

Comment:
I mean, meteor currently only edits your project constraints file if you actually write code that calls `projectConstraintsFile.addConstraints` or `projectConstraintsFile.removePackages`.  There's no other magic --- you actually have to hit code that tries to modify a thing called `projectConstraintsFile` in order for the project constraints file to be modified.  For the most part this happens either in a place where you really want the file to be edited (the add or remove commands) or on a project context that's always in a temp directory (in which case it doesn't matter much if the file is written, but it's probably helpful for debugging to write it anyway).

This flag, like the vast majority of options to the ProjectContext constructor, is an antidote for one specific place in all of Meteor where something weird and unusual happens --- we call `projectConstraintsFile.addConstraints` on a possibly non-temporary project but we don't want to change the actual file on disk.  It's weird and has a long and ugly name to alert you the fact that something strange is happening.

I don't think it's appropriate to also use this "SOMETHING WEIRD IS HAPPENING THAT OVERRIDES THE DEFAULTS" flag in cases where the defaults work fine.

Now, maybe this flag was not a great idea in the first place --- maybe this should have been an option to `addConstraints` in the one place where it's relevant.  (And, um, apparently there are two places where it was relevant and I missed one, `--get-ready`, fixing now.)  Though that would make the implementation a bit more complex (what happens if you call addConstraints once with the "don't change on disk" option and once without that option?  should it save some changes and not others?  that's what this API would suggest but it would be unused complexity).

Let me know if you have a better idea about how to design these APIs for special cases.  But with the API as designed now, passing unnecessary flags is counterproductive.